src/main/java/org/apache/cassandra/sidecar/metrics/MetricRegistryFactory.java

```
@@ -0,0 +1,116 @@
+        this.inclusions = new ArrayList<>(inclusions);
+        this.exclusions = new ArrayList<>(exclusions);
+    }
+
+    public MetricRegistry getOrCreate()
+    {
+        return getOrCreate(globalRegistryName);
+    }
+
+    public MetricRegistry getOrCreate(int cassInstanceId)
+    {
+        String instanceRegistryName = globalRegistryName + "_" + cassInstanceId;
+        return getOrCreate(instanceRegistryName);
+    }
+
+    /**
+     * Provides a {@link FilteringMetricRegistry} with given name and provided filters in configuration.
+     * @param name registry name
+     * @return a {@link MetricRegistry} that can filter out metrics
+     */
+    public MetricRegistry getOrCreate(String name)
+    {
+        // the metric registry already exists
+        if (SharedMetricRegistries.names().contains(name))
+        {
+            return SharedMetricRegistries.getOrCreate(name);
+        }
+
+        FilteringMetricRegistry metricRegistry;
+        readWriteLock.readLock().lock();
+        try
+        {
+            metricRegistry = new FilteringMetricRegistry(this::isAllowed);
+        }
+        finally
+        {
+            readWriteLock.readLock().unlock();
+        }
+        SharedMetricRegistries.add(name, metricRegistry);
+        return SharedMetricRegistries.getOrCreate(name);
+    }
+
+    /**
+     * Check if the metric is allowed to register
+     * The evaluation order is inclusions first, then exclusions. In other words,
+     * a metric name is allowed if it is in the inclusions, but not in the exclusions.
+     * <p>
+     * Note that an empty inclusions means including all
+     *
+     * @param name  metric name
+     * @return true if allowed; false otherwise
+     */
+    private boolean isAllowed(String name)
+    {
+        boolean included = inclusions.isEmpty() || inclusions.stream().anyMatch(filter -> filter.matches(name));
+        boolean excluded = exclusions.stream().anyMatch(filter -> filter.matches(name));
```

`if (included && excluded)` should we throw a configuration error?

---

In an example where configuration is 

`  include:                                   
    - type: "regex"
      value: "vertx.*"
  exclude:                                  
    - type: "regex"                         
      value: "vertx.eventbus.*"`

We want to include all vert.x metrics except the event bus metrics. In that case, both included and excluded will be true. So it not necessarily a configuration error.

---

thanks for the explanation, yeah this makes sense